src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs

```diff
@@ -180,6 +180,100 @@ public void SendFile_Synchronous(IPAddress listenAt, bool sendPreAndPostBuffers,
+            {
+                (Socket socket1, Socket socket2) = CreateConnectedSocketPair();
+                using (socket2)
+                {
+                    Task socketOperation = Task.Factory.StartNew(() =>
+                    {
+                        // Create a large file that will cause SendFile to block until the peer starts reading.
+                        string filename = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+                        using (var fs = new FileStream(filename, FileMode.CreateNew, FileAccess.Write))
+                        {
+                            fs.SetLength(20 * 1024 * 1024 /* 20MB */);
+                        }
+
+                        try
+                        {
+                            socket1.SendFile(filename);
+                        }
+                        finally
+                        {
+                            try
+                            {
+                                File.Delete(filename);
+                            }
+                            catch
+                            {}
+                        }
+                    }, TaskCreationOptions.LongRunning);
+
+                    // Wait a little so the operation is started, then Dispose.
+                    await Task.Delay(100);
+                    Task disposeTask = Task.Factory.StartNew(() =>
+                    {
+                        socket1.Dispose();
+                    }, TaskCreationOptions.LongRunning);
+
+                    Task timeoutTask = Task.Delay(30000);
+                    Assert.NotSame(timeoutTask, await Task.WhenAny(disposeTask, socketOperation, timeoutTask));
+
+                    await disposeTask;
+
+                    try
+                    {
+                        await socketOperation;
+                    }
+                    catch (SocketException se)
+                    {
+                        localSocketError = se.SocketErrorCode;
+                    }
+                    catch (ObjectDisposedException)
+                    { }
+
+                    var receiveBuffer = new byte[4096];
+                    while (true)
+                    {
+                        try
+                        {
+                            int received = socket2.Receive(receiveBuffer);
+                            if (received == 0)
+                            {
+                                break;
+                            }
+                        }
+                        catch (SocketException se)
+                        {
+                            peerSocketError = se.SocketErrorCode;
+                            break;
+                        }
+                    }
+                }
+
+                // On OSX, we're unable to unblock the on-going socket operations and
+                // perform an abortive close.
```

Comment:
Does that mean this test is going to leave blocked sockets?

Reply:
No, there are no blocked threads. The peer sees a FIN close on macOS instead of a RST close when we unblock the on-going socket operations.